pulse2percept/models/granley2021.py

```diff
@@ -867,3 +886,130 @@ def predict_percept(self, implant, t_percept=None):
             return None
         resp = self.spatial.predict_percept(implant, t_percept=t_percept)
         return resp
+
+
+
+
+class TorchBiphasicAxonMapSpatial(TorchBaseModel):
+    """ TorchBiphasicAxonMapSpatial
+
+An AxonMapModel where phosphene brightness, size, and streak length scale
+according to amplitude, frequency, and pulse duration using PyTorch.
+
+All stimuli must be BiphasicPulseTrains.
+
+This model is different than other spatial models in that it calculates
+one representative percept from all time steps of the stimulus.
+
+Brightness, size, and streak length scaling are controlled by the parameters
+bright_model, size_model, and streak model respectively. By default, these are
+set to classes that implement Eqs 3-6 from Granley 2021. These models can be
+individually customized by setting the bright_model, size_model, or streak_model
+to any python callable with signature f(freq, amp, pdur)
+
+Parameters
+----------
+bright_model: callable, optional
+    Model used to modulate percept brightness with amplitude, frequency,
+    and pulse duration
+size_model: callable, optional
+    Model used to modulate percept size with amplitude, frequency, and
+    pulse duration
+streak_model: callable, optional
+    Model used to modulate percept streak length with amplitude, frequency,
+    and pulse duration
+do_thresholding: boolean
+    Use probabilistic sigmoid thresholding, default: False
+**params: dict, optional
+    Additional parameters to customize the model
+"""
+    def __init__(self, p2pmodel):
+        super().__init__(p2pmodel)
+        
+        self.rho = torch.tensor(p2pmodel.rho, device=self.device)
+        self.shape = p2pmodel.grid.shape
+        self.axon_contrib = torch.tensor(p2pmodel.axon_contrib, device=self.device)
+        self.rho = torch.tensor(p2pmodel.rho, device=self.device)
+        self.axlambda = torch.tensor(p2pmodel.axlambda, device=self.device)
+            
+        self.a0 = torch.tensor(p2pmodel.a0, device=self.device)
+        self.a1 = torch.tensor(p2pmodel.a1, device=self.device)
+        self.a2 = torch.tensor(p2pmodel.a2, device=self.device)
+        self.a3 = torch.tensor(p2pmodel.a3, device=self.device)
+        self.a4 = torch.tensor(p2pmodel.a4, device=self.device)
+        self.a5 = torch.tensor(p2pmodel.a5, device=self.device)
+        self.a6 = torch.tensor(p2pmodel.a6, device=self.device)
+        self.a7 = torch.tensor(p2pmodel.a7, device=self.device)
+        self.a8 = torch.tensor(p2pmodel.a8, device=self.device)
+        self.a9 = torch.tensor(p2pmodel.a9, device=self.device)
+        
+        self.locs = {}
+        # Other parameters from initialization args
+        self.percept_shape = p2pmodel.grid.shape  # p2pmodel has a grid attribute defining percept shape
+        self.thresh_percept = p2pmodel.thresh_percept
+    
+    
+    def forward(self, stims, e_locs, model_params=None):
+        if model_params is None:
+            rho = self.rho
```

Comment:
Can we do these all the same way? The one liners you have for a0-a9 would be great here too